content/blog/2018/08/2018-08-14-code-coverage-api-plugin-1.0-release.adoc

```diff
@@ -0,0 +1,205 @@
+
+==== Parallel pipeline support
+
+We support parallel pipeline. You can call Code Coverage API plugin in different branches like that:
+[source, groovy]
+----
+node {
+    parallel firstBranch: {
+        publishCoverage adapters: [jacoco('target/site/jacoco/jacoco.xml')]    
+}, secondBranch: {
+        publishCoverage adapters: [jacoco('jacoco.xml')]
+    }
+}
+----
+
+==== Reports combining
+You can add tag on `publishCoverage` and Code Coverage API plugin will combine reports have same tag
+
+----
+node {
+    parallel firstBranch: {
+        publishCoverage adapters: [jacoco('target/site/jacoco/jacoco.xml')], tag: ‘t’
+}, secondBranch: {
+        publishCoverage adapters: [jacoco('jacoco.xml')], tag: ‘t’
+    }
+}
+----
+
+==== Rest API
+We provide REST API for others to retrieve coverage data:
+
+* Coverage result: `.../\{buildNumber\}/coverage/.../result/api/\{json|xml\}`
+* Trend result: `.../\{buildNumber\}/coverage/.../trend/api/\{json|xml\}`
+* Coverage result of last build: `.../\{buildNumber\}/coverage/.../last/result/api/\{json|xml\}`
+* Trend result of last build: `.../\{buildNumber\}/coverage/.../last/trend/api/\{json|xml\}`
+
+==== Failed conditions and flexible threshold setting
+You can set different failed conditions and threholds to control build result.
+image:/images/post-images/code-coverage-api/thresholds-config.png[title="Thresholds Configuration" role="center"]
+
+If the thresholds satisfy the failed conditions, it will make this build failed. 
+image:/images/post-images/code-coverage-api/thresholds-result.png[title="Thresholds Result" role="center"]
+
+==== Other small features
+
+We also have other small features like auto detecting reports, coverage filters, etc. You can find more information about these features in the plugin documentation.
+
+=== Architecture
+
+This API plugin will mainly do these things:
+
+* Find coverage reports according to the user’s config.
+* Use adapters to convert reports into the our standard format.
+* Parse standard format reports, and aggregate them.
+* Show parsed result in a chart.
+
+So, we can implement code coverage publishing by simply writing an adapter, and such adapter only needs to do one thing - convert a coverage report into the standard format. The implementation is based on extension points, so new adapters can be created in separate plugins. In order to simplify conversion for XML reports, there is also an abstraction layer which allows creating XSLT-based converters.
+
+*The below diagram show the architecture of Code Coverage API plugin*
+
+image:/images/post-images/code-coverage-api/architecture.png[title="Code Coverage API Plugin Architecture" role="center"]
+
+=== LLVM Coverage Plugin
+TODO
+
+=== How to Implement It
+
+We can implement a coverage tool by implementing CoverageReportAdapter extension point. For example, by using the provided abstract layer, we can implement JaCoCo simple like this:
```

Comment:
This is really about implementing a new plugin, right? It might be better to use the word `plugin` rather than `tool`